batch/batch/worker.py

```diff
@@ -81,30 +84,94 @@ def free(self, port):
+                else:
+                    raise
+            except aiohttp.client_exceptions.ServerDisconnectedError:
+                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
+            except asyncio.TimeoutError:
+                local_timeout *= 2
+                log.exception(f'in docker call to {f.__name__} for {name}, retrying with timeout {local_timeout} seconds', stack_info=True)
+            # exponentially back off, up to (expected) max of 30s
+            t = delay * random.random()
+            await asyncio.sleep(t)
+            delay = min(delay * 2, 60.0)
+    return _wrapper
+
+
+async def create_container(config, name):
+    error = 0
+    delay = 1
     while True:
         try:
-            return await f(*args, **kwargs)
+            return await docker.containers.create(config, name=name)
         except DockerError as e:
-            # 408 request timeout, 503 service unavailable
-            if e.status == 408 or e.status == 503:
-                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-            # DockerError(500, 'Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
-            # DockerError(500, 'error creating overlay mount to /var/lib/docker/overlay2/545a1337742e0292d9ed197b06fe900146c85ab06e468843cd0461c3f34df50d/merged: device or resource busy'
-            elif e.status == 500 and ("request canceled while waiting for connection" in e.message or
-                                      re.match("error creating overlay mount.*device or resource busy", e.message)):
-                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-            else:
+            # exponentially back off, up to (expected) max of 30s
+            t = delay * random.random()
+            await asyncio.sleep(t)
+            delay = min(delay * 2, 60.0)
+
+            error += 1
+            if error == 5:
+                log.exception(f'encountered 5 errors while creating container {name}, aborting', stack_info=True)
                 raise
-        except aiohttp.client_exceptions.ServerDisconnectedError:
-            log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-        except asyncio.TimeoutError:
-            log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-        # exponentially back off, up to (expected) max of 30s
-        t = delay * random.random()
-        await asyncio.sleep(t)
-        delay = min(delay * 2, 60.0)
+
+            # 409 container with name already exists
+            if e.status == 409:
+                try:
+                    return await docker.containers.get(name)
+                except DockerError as eget:
+                    # 404 No such container
+                    if eget.status == 404:
+                        continue
+                    raise
```

Comment:
I'm inclined to leave this raise off and let everything fall through the same default raise case.